File: vox/rust/roam-postcard/src/plan.rs
use std::collections::HashMap;

use facet_core::{Shape, Type, UserType};
use roam_types::{
    ExtractedSchemas, FieldSchema, Schema, SchemaHash, SchemaKind, SchemaRegistry, TypeRef,
    VariantPayload, VariantSchema,
};

use crate::error::{PathSegment, SchemaSide, TranslationError, TranslationErrorKind};

/// A precomputed plan for deserializing postcard bytes into a local type.
///
/// This is a recursive enum that mirrors the shape of data. Each variant
/// carries sub-plans for its children, so translation works through
/// arbitrarily nested containers (Vec, Option, Map, etc.).
#[derive(Debug)]
pub enum TranslationPlan {
    /// Identity — no translation needed. Used for leaves (scalars, etc.)
    /// and for container elements when remote and local types match.
    Identity,

    /// Struct or tuple-struct: field reordering, skipping unknown, filling defaults.
    Struct {
        /// One op per remote field, in remote wire order.
        field_ops: Vec<FieldOp>,
        /// Nested plans for fields that need translation, keyed by local field index.
        nested: HashMap<usize, TranslationPlan>,
    },

    /// Enum: variant remapping + per-variant field plans.
    Enum {
        /// For each remote variant index, the local variant index (or None if unknown).
        variant_map: Vec<Option<usize>>,
        /// Per-variant field plans, keyed by remote variant index.
        variant_plans: HashMap<usize, TranslationPlan>,
        /// Nested plans for newtype/tuple variant inner types, keyed by local variant index.
        nested: HashMap<usize, TranslationPlan>,
    },

    /// Tuple: positional elements with nested plans.
    Tuple {
        field_ops: Vec<FieldOp>,
        nested: HashMap<usize, TranslationPlan>,
    },

    /// List/Vec/Slice: element type needs translation.
    List { element: Box<TranslationPlan> },

    /// Option: inner type needs translation.
    Option { inner: Box<TranslationPlan> },

    /// Map: key and/or value types need translation.
    Map {
        key: Box<TranslationPlan>,
        value: Box<TranslationPlan>,
    },

    /// Fixed-size array: element type needs translation.
    Array { element: Box<TranslationPlan> },

    /// Pointer (Box, Arc, etc.): pointee needs translation.
    Pointer { pointee: Box<TranslationPlan> },
}

#[derive(Debug)]
pub enum FieldOp {
    /// Read this remote field into local field at `local_index`.
    Read { local_index: usize },
    /// Skip this remote field (not present in local type).
    Skip { type_ref: TypeRef },
}

/// A schema set: the root schema (with Vars resolved) + the registry.
#[derive(Debug)]
pub struct SchemaSet {
    pub root: Schema,
    pub registry: SchemaRegistry,
}

impl SchemaSet {
    /// Build a SchemaSet from a raw list of schemas (e.g. received from the wire).
    /// The root is the last schema. Its kind is used as-is (no Var resolution).
    pub fn from_schemas(schemas: Vec<Schema>) -> Self {
        let root = schemas.last().cloned().expect("empty schema list");
        let registry = roam_types::build_registry(&schemas);
        SchemaSet { root, registry }
    }

    /// Build a SchemaSet from extracted schemas.
    /// The root TypeRef is used to resolve any Var references in the root schema.
    pub fn from_extracted(extracted: ExtractedSchemas) -> Self {
        let registry = roam_types::build_registry(&extracted.schemas);
        // Resolve the root schema's kind using the root TypeRef's args.
        let root_kind = extracted
            .root_type_ref
            .resolve_kind(&registry)
            .expect("root schema must be in registry");
        let root_id = match &extracted.root_type_ref {
            TypeRef::Concrete { type_id, .. } => *type_id,
            TypeRef::Var { .. } => unreachable!("root type ref is never a Var"),
        };
        let root = Schema {
            id: root_id,
            type_params: vec![],
            kind: root_kind,
        };
        SchemaSet { root, registry }
    }
}

/// Input to `build_plan`: identifies which side is remote and which is local.
pub struct PlanInput<'a> {
    pub remote: &'a SchemaSet,
    pub local: &'a SchemaSet,
}

/// Build the trivial identity plan from a local Shape alone.
/// Every field maps 1:1, no skips, no defaults. Used when no remote schema
/// is available (same types on both sides).
pub fn build_identity_plan(shape: &'static Shape) -> TranslationPlan {
    match shape.ty {
        Type::User(UserType::Struct(struct_type)) => {
            let field_ops = (0..struct_type.fields.len())
                .map(|i| FieldOp::Read { local_index: i })
                .collect();
            TranslationPlan::Struct {
                field_ops,
                nested: HashMap::new(),
            }
        }
        Type::User(UserType::Enum(enum_type)) => {
            let variant_map = (0..enum_type.variants.len()).map(Some).collect();
            let mut variant_plans = HashMap::new();
            for (i, variant) in enum_type.variants.iter().enumerate() {
                let field_ops = (0..variant.data.fields.len())
                    .map(|j| FieldOp::Read { local_index: j })
                    .collect();
                variant_plans.insert(
                    i,
                    TranslationPlan::Struct {
                        field_ops,
                        nested: HashMap::new(),
                    },
                );
            }
            TranslationPlan::Enum {
                variant_map,
                variant_plans,
                nested: HashMap::new(),
            }
        }
        _ => TranslationPlan::Identity,
    }
}

/// Build a translation plan by comparing remote and local schemas.
///
/// Both sides are represented as schemas — the same extraction logic
/// (channel unwrapping, transparent wrappers, etc.) has already run on
/// both. This avoids mismatches between schema representation and raw
/// Shape inspection.
// r[impl schema.translation.field-matching]
// r[impl schema.translation.skip-unknown]
// r[impl schema.translation.fill-defaults]
// r[impl schema.translation.reorder]
// r[impl schema.errors.early-detection]
pub fn build_plan(input: &PlanInput) -> Result<TranslationPlan, TranslationError> {
    let remote = &input.remote.root;
    let local = &input.local.root;

    // Validate type names match for nominal types (struct/enum).
    if let (Some(remote_name), Some(local_name)) = (remote.name(), local.name())
        && remote_name != local_name
    {
        return Err(TranslationError::new(TranslationErrorKind::NameMismatch {
            remote: remote.clone(),
            local: local.clone(),
        }));
    }

    match (&remote.kind, &local.kind) {
        (
            SchemaKind::Struct {
                fields: remote_fields,
                ..
            },
            SchemaKind::Struct {
                fields: local_fields,
                ..
            },
        ) => build_struct_plan(remote_fields, local_fields, remote, local, input),
        (
            SchemaKind::Enum {
                variants: remote_variants,
                ..
            },
            SchemaKind::Enum {
                variants: local_variants,
                ..
            },
        ) => build_enum_plan(remote_variants, local_variants, remote, local, input),
        (
            SchemaKind::Tuple {
                elements: remote_elements,
            },
            SchemaKind::Tuple {
                elements: local_elements,
            },
        ) => build_tuple_plan(remote_elements, local_elements, remote, local, input),
        // Container types — recurse into element/value types
        (
            SchemaKind::List {
                element: remote_elem,
            },
            SchemaKind::List {
                element: local_elem,
            },
        ) => {
            let element_plan = nested_plan(remote_elem, local_elem, input)?;
            Ok(TranslationPlan::List {
                element: Box::new(element_plan.unwrap_or(TranslationPlan::Identity)),
            })
        }
        (
            SchemaKind::Option {
                element: remote_elem,
            },
            SchemaKind::Option {
                element: local_elem,
            },
        ) => {
            let inner_plan = nested_plan(remote_elem, local_elem, input)?;
            Ok(TranslationPlan::Option {
                inner: Box::new(inner_plan.unwrap_or(TranslationPlan::Identity)),
            })
        }
        (
            SchemaKind::Map {
                key: remote_key,
                value: remote_val,
            },
            SchemaKind::Map {
                key: local_key,
                value: local_val,
            },
        ) => {
            let key_plan = nested_plan(remote_key, local_key, input)?;
            let val_plan = nested_plan(remote_val, local_val, input)?;
            Ok(TranslationPlan::Map {
                key: Box::new(key_plan.unwrap_or(TranslationPlan::Identity)),
                value: Box::new(val_plan.unwrap_or(TranslationPlan::Identity)),
            })
        }
        (
            SchemaKind::Array {
                element: remote_elem,
                ..
            },
            SchemaKind::Array {
                element: local_elem,
                ..
            },
        ) => {
            let element_plan = nested_plan(remote_elem, local_elem, input)?;
            Ok(TranslationPlan::Array {
                element: Box::new(element_plan.unwrap_or(TranslationPlan::Identity)),
            })
        }
        // Primitives — no translation needed
        (SchemaKind::Primitive { .. }, SchemaKind::Primitive { .. }) => {
            Ok(TranslationPlan::Identity)
        }
        // Kind mismatch
        _ => Err(TranslationError::new(TranslationErrorKind::KindMismatch {
            remote: remote.clone(),
            local: local.clone(),
        })),
    }
}

/// Build a nested plan for two TypeRefs looked up in their respective registries.
/// Handles generic types by resolving Var references using the TypeRef's args.
fn nested_plan(
    remote_type_ref: &TypeRef,
    local_type_ref: &TypeRef,
    input: &PlanInput,
) -> Result<Option<TranslationPlan>, TranslationError> {
    let resolve_schema = |type_ref: &TypeRef, registry: &SchemaRegistry, side: SchemaSide| {
        let type_id = match type_ref {
            TypeRef::Concrete { type_id, .. } => *type_id,
            TypeRef::Var { name } => {
                return Err(TranslationError::new(TranslationErrorKind::UnresolvedVar {
                    name: format!("{name:?}"),
                    side,
                }));
            }
        };
        let kind = type_ref.resolve_kind(registry).ok_or_else(|| {
            TranslationError::new(TranslationErrorKind::SchemaNotFound { type_id, side })
        })?;
        let base = registry.get(&type_id).ok_or_else(|| {
            TranslationError::new(TranslationErrorKind::SchemaNotFound { type_id, side })
        })?;
        Ok(Schema {
            id: base.id,
            type_params: vec![],
            kind,
        })
    };

    let remote_schema =
        resolve_schema(remote_type_ref, &input.remote.registry, SchemaSide::Remote)?;
    let local_schema = resolve_schema(local_type_ref, &input.local.registry, SchemaSide::Local)?;

    let sub_input = PlanInput {
        remote: &SchemaSet {
            root: remote_schema,
            registry: input.remote.registry.clone(),
        },
        local: &SchemaSet {
            root: local_schema,
            registry: input.local.registry.clone(),
        },
    };
    build_plan(&sub_input).map(Some)
}

fn build_struct_plan(
    remote_fields: &[FieldSchema],
    local_fields: &[FieldSchema],
    remote_schema: &Schema,
    _local_schema: &Schema,
    input: &PlanInput,
) -> Result<TranslationPlan, TranslationError> {
    let mut field_ops = Vec::with_capacity(remote_fields.len());
    let mut nested = HashMap::new();
    let mut matched_local = vec![false; local_fields.len()];

    for remote_field in remote_fields {
        if let Some((local_idx, local_field)) = local_fields
            .iter()
            .enumerate()
            .find(|(_, f)| f.name == remote_field.name)
        {
            matched_local[local_idx] = true;
            field_ops.push(FieldOp::Read {
                local_index: local_idx,
            });

            // r[impl schema.translation.type-compat]
            let nested_plan = nested_plan(&remote_field.type_ref, &local_field.type_ref, input)
                .map_err(|e| e.with_path_prefix(PathSegment::Field(remote_field.name.clone())))?;
            if let Some(plan) = nested_plan {
                nested.insert(local_idx, plan);
            }
        } else {
            field_ops.push(FieldOp::Skip {
                type_ref: remote_field.type_ref.clone(),
            });
        }
    }

    // r[impl schema.errors.missing-required]
    for (i, matched) in matched_local.iter().enumerate() {
        if !matched && local_fields[i].required {
            return Err(TranslationError::new(
                TranslationErrorKind::MissingRequiredField {
                    field: local_fields[i].clone(),
                    remote_struct: remote_schema.clone(),
                },
            ));
        }
    }

    Ok(TranslationPlan::Struct { field_ops, nested })
}

fn build_tuple_plan(
    remote_elements: &[TypeRef<SchemaHash>],
    local_elements: &[TypeRef<SchemaHash>],
    remote_schema: &Schema,
    local_schema: &Schema,
    input: &PlanInput,
) -> Result<TranslationPlan, TranslationError> {
    if remote_elements.len() != local_elements.len() {
        return Err(TranslationError::new(
            TranslationErrorKind::TupleLengthMismatch {
                remote: remote_schema.clone(),
                local: local_schema.clone(),
                remote_len: remote_elements.len(),
                local_len: local_elements.len(),
            },
        ));
    }

    let mut field_ops = Vec::with_capacity(remote_elements.len());
    let mut nested = HashMap::new();

    for (i, (remote_elem, local_elem)) in remote_elements
        .iter()
        .zip(local_elements.iter())
        .enumerate()
    {
        field_ops.push(FieldOp::Read { local_index: i });

        let nested_plan = nested_plan(remote_elem, local_elem, input)
            .map_err(|e| e.with_path_prefix(PathSegment::Index(i)))?;
        if let Some(plan) = nested_plan {
            nested.insert(i, plan);
        }
    }

    Ok(TranslationPlan::Tuple { field_ops, nested })
}

// r[impl schema.translation.enum]
// r[impl schema.translation.enum.missing-variant]
// r[impl schema.translation.enum.payload-compat]
fn build_enum_plan(
    remote_variants: &[VariantSchema],
    local_variants: &[VariantSchema],
    _remote_schema: &Schema,
    _local_schema: &Schema,
    input: &PlanInput,
) -> Result<TranslationPlan, TranslationError> {
    let mut variant_map = Vec::with_capacity(remote_variants.len());
    let mut variant_plans = HashMap::new();
    let mut nested = HashMap::new();

    for (remote_idx, remote_variant) in remote_variants.iter().enumerate() {
        if let Some((local_idx, local_variant)) = local_variants
            .iter()
            .enumerate()
            .find(|(_, v)| v.name == remote_variant.name)
        {
            variant_map.push(Some(local_idx));

            match (&remote_variant.payload, &local_variant.payload) {
                // Both struct variants — build a per-variant field plan
                (
                    VariantPayload::Struct {
                        fields: remote_fields,
                    },
                    VariantPayload::Struct {
                        fields: local_fields,
                    },
                ) => {
                    let variant_field_ops: Vec<FieldOp> = remote_fields
                        .iter()
                        .map(|rf| {
                            if let Some((local_field_idx, _)) = local_fields
                                .iter()
                                .enumerate()
                                .find(|(_, f)| f.name == rf.name)
                            {
                                FieldOp::Read {
                                    local_index: local_field_idx,
                                }
                            } else {
                                FieldOp::Skip {
                                    type_ref: rf.type_ref.clone(),
                                }
                            }
                        })
                        .collect();
                    variant_plans.insert(
                        remote_idx,
                        TranslationPlan::Struct {
                            field_ops: variant_field_ops,
                            nested: HashMap::new(),
                        },
                    );
                }
                // Both newtype — build a nested plan for the inner type
                (
                    VariantPayload::Newtype {
                        type_ref: remote_inner_ref,
                    },
                    VariantPayload::Newtype {
                        type_ref: local_inner_ref,
                    },
                ) => {
                    let inner_plan = nested_plan(remote_inner_ref, local_inner_ref, input)
                        .map_err(|e| {
                            e.with_path_prefix(PathSegment::Variant(remote_variant.name.clone()))
                        })?;
                    if let Some(plan) = inner_plan {
                        nested.insert(local_idx, plan);
                    }
                }
                // Both tuple — check arity matches and build nested plans
                (
                    VariantPayload::Tuple {
                        types: remote_types,
                    },
                    VariantPayload::Tuple { types: local_types },
                ) => {
                    let tuple_plan = build_tuple_plan(
                        remote_types,
                        local_types,
                        _remote_schema,
                        _local_schema,
                        input,
                    )
                    .map_err(|e| {
                        e.with_path_prefix(PathSegment::Variant(remote_variant.name.clone()))
                    })?;
                    variant_plans.insert(remote_idx, tuple_plan);
                }
                (VariantPayload::Unit, VariantPayload::Unit) => {}
                // Payload kind mismatch within a variant
                _ => {
                    return Err(TranslationError::new(
                        TranslationErrorKind::IncompatibleVariantPayload {
                            remote_variant: remote_variant.clone(),
                            local_variant: local_variant.clone(),
                        },
                    )
                    .with_path_prefix(PathSegment::Variant(remote_variant.name.clone())));
                }
            }
        } else {
            // r[impl schema.translation.enum.unknown-variant]
            variant_map.push(None);
        }
    }

    Ok(TranslationPlan::Enum {
        variant_map,
        variant_plans,
        nested,
    })
}

File: vox/typescript/packages/roam-postcard/src/plan.ts
// Translation plan for schema evolution.
//
// Ported from Rust's roam-postcard/src/plan.rs. Compares remote and local
// schemas and produces a plan that drives single-pass postcard decoding with
// field reordering, skipping, and default-filling.

import type {
  Schema,
  SchemaKind,
  SchemaRegistry,
  TypeRef,
  FieldSchema,
  VariantSchema,
  VariantPayload,
  SchemaHash,
} from "./schema.ts";
import { resolveTypeRef } from "./schema.ts";

// ============================================================================
// TranslationPlan
// ============================================================================

export type TranslationPlan =
  | { tag: "identity" }
  | { tag: "struct"; field_ops: FieldOp[]; nested: Map<number, TranslationPlan> }
  | {
      tag: "enum";
      variant_map: (number | null)[];
      variant_plans: Map<number, TranslationPlan>;
      nested: Map<number, TranslationPlan>;
    }
  | { tag: "tuple"; field_ops: FieldOp[]; nested: Map<number, TranslationPlan> }
  | { tag: "list"; element: TranslationPlan }
  | { tag: "option"; inner: TranslationPlan }
  | { tag: "map"; key: TranslationPlan; value: TranslationPlan }
  | { tag: "array"; element: TranslationPlan }
  | { tag: "pointer"; pointee: TranslationPlan };

export type FieldOp =
  | { tag: "read"; local_index: number }
  | { tag: "skip"; type_ref: TypeRef };

export const IDENTITY: TranslationPlan = { tag: "identity" };

// ============================================================================
// SchemaSet
// ============================================================================

export interface SchemaSet {
  root: Schema;
  registry: SchemaRegistry;
}

/** Build a SchemaSet from a list of schemas (e.g. received from the wire). */
export function schemaSetFromSchemas(schemas: Schema[]): SchemaSet {
  const root = schemas[schemas.length - 1];
  if (!root) throw new Error("empty schema list");
  const registry: SchemaRegistry = new Map();
  for (const s of schemas) {
    registry.set(s.id, s);
  }
  return { root, registry };
}

// ============================================================================
// buildPlan
// ============================================================================

export class TranslationError extends Error {
  path: string[];

  constructor(message: string) {
    super(message);
    this.name = "TranslationError";
    this.path = [];
  }

  withPathPrefix(segment: string): TranslationError {
    this.path.unshift(segment);
    this.message = `${this.path.join(".")}: ${this.message}`;
    return this;
  }
}

/**
 * Build a translation plan by comparing remote and local schemas.
 *
 * Returns `null` if the schemas are identical (identity plan).
 */
export function buildPlan(remote: SchemaSet, local: SchemaSet): TranslationPlan {
  return buildPlanInner(remote.root, local.root, remote.registry, local.registry);
}

function buildPlanInner(
  remote: Schema,
  local: Schema,
  remoteReg: SchemaRegistry,
  localReg: SchemaRegistry,
): TranslationPlan {
  // Validate type names match for nominal types
  const remoteName = schemaName(remote.kind);
  const localName = schemaName(local.kind);
  if (remoteName && localName && remoteName !== localName) {
    throw new TranslationError(
      `type name mismatch: remote "${remoteName}" vs local "${localName}"`,
    );
  }

  const rk = remote.kind;
  const lk = local.kind;

  if (isByteBufferKind(rk, remoteReg) && isByteBufferKind(lk, localReg)) {
    return IDENTITY;
  }

  if (rk.tag !== lk.tag) {
    throw new TranslationError(
      `schema kind mismatch: remote "${rk.tag}" vs local "${lk.tag}"`,
    );
  }

  switch (rk.tag) {
    case "struct": {
      const lkStruct = lk as Extract<SchemaKind, { tag: "struct" }>;
      return buildStructPlan(rk.fields, lkStruct.fields, remote, remoteReg, localReg);
    }
    case "enum": {
      const lkEnum = lk as Extract<SchemaKind, { tag: "enum" }>;
      return buildEnumPlan(rk.variants, lkEnum.variants, remote, local, remoteReg, localReg);
    }
    case "tuple": {
      const lkTuple = lk as Extract<SchemaKind, { tag: "tuple" }>;
      return buildTuplePlan(rk.elements, lkTuple.elements, remote, local, remoteReg, localReg);
    }
    case "list": {
      const lkList = lk as Extract<SchemaKind, { tag: "list" }>;
      const element = nestedPlan(rk.element, lkList.element, remoteReg, localReg);
      return { tag: "list", element: element ?? IDENTITY };
    }
    case "option": {
      const lkOpt = lk as Extract<SchemaKind, { tag: "option" }>;
      const inner = nestedPlan(rk.element, lkOpt.element, remoteReg, localReg);
      return { tag: "option", inner: inner ?? IDENTITY };
    }
    case "map": {
      const lkMap = lk as Extract<SchemaKind, { tag: "map" }>;
      const key = nestedPlan(rk.key, lkMap.key, remoteReg, localReg);
      const value = nestedPlan(rk.value, lkMap.value, remoteReg, localReg);
      return { tag: "map", key: key ?? IDENTITY, value: value ?? IDENTITY };
    }
    case "array": {
      const lkArr = lk as Extract<SchemaKind, { tag: "array" }>;
      const element = nestedPlan(rk.element, lkArr.element, remoteReg, localReg);
      return { tag: "array", element: element ?? IDENTITY };
    }
    case "primitive":
      {
        const lkPrimitive = lk as Extract<SchemaKind, { tag: "primitive" }>;
        if (rk.primitive_type !== lkPrimitive.primitive_type) {
          throw new TranslationError(
            `primitive type mismatch: remote "${rk.primitive_type}" vs local "${lkPrimitive.primitive_type}"`,
          );
        }
      }
      return IDENTITY;
    case "channel":
      return IDENTITY;
  }
}

function nestedPlan(
  remoteRef: TypeRef,
  localRef: TypeRef,
  remoteReg: SchemaRegistry,
  localReg: SchemaRegistry,
): TranslationPlan | null {
  const resolveSchema = (
    ref_: TypeRef,
    registry: SchemaRegistry,
    side: string,
  ): Schema => {
    if (ref_.tag === "var") {
      throw new TranslationError(`unresolved type variable "${ref_.name}" on ${side} side`);
    }
    const kind = resolveTypeRef(ref_, registry);
    if (!kind) {
      throw new TranslationError(
        `schema not found for type_id ${ref_.type_id} on ${side} side`,
      );
    }
    const base = registry.get(ref_.type_id);
    if (!base) {
      throw new TranslationError(
        `schema not found for type_id ${ref_.type_id} on ${side} side`,
      );
    }
    return { id: base.id, type_params: [], kind };
  };

  const remoteSchema = resolveSchema(remoteRef, remoteReg, "remote");
  const localSchema = resolveSchema(localRef, localReg, "local");

  return buildPlanInner(remoteSchema, localSchema, remoteReg, localReg);
}

function buildStructPlan(
  remoteFields: FieldSchema[],
  localFields: FieldSchema[],
  remoteSchema: Schema,
  remoteReg: SchemaRegistry,
  localReg: SchemaRegistry,
): TranslationPlan {
  const fieldOps: FieldOp[] = [];
  const nested = new Map<number, TranslationPlan>();
  const matched = new Array(localFields.length).fill(false);

  for (const rf of remoteFields) {
    const localIdx = localFields.findIndex((f) => f.name === rf.name);
    if (localIdx >= 0) {
      matched[localIdx] = true;
      fieldOps.push({ tag: "read", local_index: localIdx });

      const np = nestedPlan(rf.type_ref, localFields[localIdx].type_ref, remoteReg, localReg);
      if (np) nested.set(localIdx, np);
    } else {
      fieldOps.push({ tag: "skip", type_ref: rf.type_ref });
    }
  }

  // Check for missing required fields
  for (let i = 0; i < localFields.length; i++) {
    if (!matched[i] && !fieldHasDefault(localFields[i], localReg)) {
      throw new TranslationError(
        `required field "${localFields[i].name}" missing from remote schema "${schemaName(remoteSchema.kind) ?? "?"}"`,
      );
    }
  }

  return { tag: "struct", field_ops: fieldOps, nested };
}

function fieldHasDefault(
  field: FieldSchema,
  registry: SchemaRegistry,
): boolean {
  if (!field.required) {
    return true;
  }
  const kind = resolveTypeRef(field.type_ref, registry);
  return kind?.tag === "option";
}

function buildTuplePlan(
  remoteElements: TypeRef[],
  localElements: TypeRef[],
  _remoteSchema: Schema,
  _localSchema: Schema,
  remoteReg: SchemaRegistry,
  localReg: SchemaRegistry,
): TranslationPlan {
  if (remoteElements.length !== localElements.length) {
    throw new TranslationError(
      `tuple length mismatch: remote ${remoteElements.length} vs local ${localElements.length}`,
    );
  }

  const fieldOps: FieldOp[] = [];
  const nested = new Map<number, TranslationPlan>();

  for (let i = 0; i < remoteElements.length; i++) {
    fieldOps.push({ tag: "read", local_index: i });
    const np = nestedPlan(remoteElements[i], localElements[i], remoteReg, localReg);
    if (np) nested.set(i, np);
  }

  return { tag: "tuple", field_ops: fieldOps, nested };
}

function buildEnumPlan(
  remoteVariants: VariantSchema[],
  localVariants: VariantSchema[],
  remoteSchema: Schema,
  localSchema: Schema,
  remoteReg: SchemaRegistry,
  localReg: SchemaRegistry,
): TranslationPlan {
  const variantMap: (number | null)[] = [];
  const variantPlans = new Map<number, TranslationPlan>();
  const nested = new Map<number, TranslationPlan>();

  for (let remoteIdx = 0; remoteIdx < remoteVariants.length; remoteIdx++) {
    const rv = remoteVariants[remoteIdx];
    const localIdx = localVariants.findIndex((v) => v.name === rv.name);

    if (localIdx < 0) {
      // Unknown remote variant
      variantMap.push(null);
      continue;
    }

    variantMap.push(localIdx);
    const lv = localVariants[localIdx];

    if (rv.payload.tag !== lv.payload.tag) {
      throw new TranslationError(
        `variant "${rv.name}": payload kind mismatch "${rv.payload.tag}" vs "${lv.payload.tag}"`,
      );
    }

    switch (rv.payload.tag) {
      case "struct": {
        const lvPayload = lv.payload as Extract<VariantPayload, { tag: "struct" }>;
        const varFieldOps: FieldOp[] = rv.payload.fields.map((rf) => {
          const idx = lvPayload.fields.findIndex((f) => f.name === rf.name);
          if (idx >= 0) {
            return { tag: "read" as const, local_index: idx };
          } else {
            return { tag: "skip" as const, type_ref: rf.type_ref };
          }
        });
        variantPlans.set(remoteIdx, {
          tag: "struct",
          field_ops: varFieldOps,
          nested: new Map(),
        });
        break;
      }
      case "newtype": {
        const lvPayload = lv.payload as Extract<VariantPayload, { tag: "newtype" }>;
        const np = nestedPlan(rv.payload.type_ref, lvPayload.type_ref, remoteReg, localReg);
        if (np) nested.set(localIdx, np);
        break;
      }
      case "tuple": {
        const lvPayload = lv.payload as Extract<VariantPayload, { tag: "tuple" }>;
        const tuplePlan = buildTuplePlan(
          rv.payload.types,
          lvPayload.types,
          remoteSchema,
          localSchema,
          remoteReg,
          localReg,
        );
        variantPlans.set(remoteIdx, tuplePlan);
        break;
      }
      case "unit":
        break;
    }
  }

  return { tag: "enum", variant_map: variantMap, variant_plans: variantPlans, nested };
}

function schemaName(kind: SchemaKind): string | null {
  if (kind.tag === "struct" || kind.tag === "enum") return kind.name;
  return null;
}

function isByteBufferKind(
  kind: SchemaKind,
  registry: SchemaRegistry,
): boolean {
  if (kind.tag === "primitive") {
    return kind.primitive_type === "bytes";
  }
  if (kind.tag !== "list") {
    return false;
  }
  const elementKind = resolveTypeRef(kind.element, registry);
  return elementKind?.tag === "primitive" && elementKind.primitive_type === "u8";
}
